### PR TITLE
Register new client device if token's registrationExpiresAt field is …

### PR DESCRIPTION
**Pull request by peterrsongg**
…null

This is to provide backwards compatibility as mentioned in [this pr] (https://github.com/aws/aws-sdk-net/pull/3083). Some systems may generate tokens where the `RegistrationExpiresAt` field is null so in that case we should just say that the clientRegistration is expired and generate a new token.

## Description
If the `registrationExpiresAt` field is null or empty, treat the token as expired and generate a new token.

## Motivation and Context
Solves backwards compatibility issues, where a null RegistrationExpiresAt was previously valid.

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist

- [ ] My code follows the code style of this project
- [ ] My change requires a change to the documentation
- [ ] I have updated the documentation accordingly
- [ ] I have read the **README** document
- [ ] I have added tests to cover my changes
- [ x] All new and existing tests passed

## License

- [x ] I confirm that this pull request can be released under the Apache 2 license

[issues]: https://github.com/aws/aws-sdk-net/issues
[license]: http://aws.amazon.com/apache2.0/
[cla]: http://en.wikipedia.org/wiki/Contributor_License_Agreement

### CODE REVIEW

File: sdk/src/Core/Amazon.Runtime/Credentials/Internal/_bcl45+netstandard/SsoTokenUtils.cs [hunk 1]
```diff
@@ -99,13 +99,17 @@ public static string AsJson(this SsoToken token)
...
+        /// In these cases we treat the client registration as expired and generate a new token.
         /// </summary>
         /// <param name="token">The sso token</param>
         /// <returns>This returns true if <seealso cref="SsoToken.RegistrationExpiresAt"/> is within 5 minutes of expiration. False otherwise</returns>
         public static bool RegisteredClientExpired(this SsoToken token)
         {
             if (null == token)
                 throw new ArgumentNullException(nameof(token));
+            if (string.IsNullOrEmpty(token.RegistrationExpiresAt))
+                    return true;
```
Comment (ashovlin) [inline sdk/src/Core/Amazon.Runtime/Credentials/Internal/_bcl45+netstandard/SsoTokenUtils.cs:112]: I haven't actually used this, does the user experience differ for whether we're calling `GenerateNewTokenAsync` vs. `RefreshTokenAsync`?

Still debating the two proposals from @awschristou in https://github.com/aws/aws-sdk-net/pull/3083/files#r1389714680.

If using the SDK with a tool that never sets `RegistrationExpiresAt` originally, will we be forcing new tokens to eagerly even if the registration had a lot of time left?

**Review by ashovlin — COMMENTED** [review]
Can we add a test for the case where `RegistrationExpiresAt` is null if we don't have one? Needing to fix this so soon suggests there may be a gap.

File: sdk/src/Core/Amazon.Runtime/Credentials/Internal/_bcl45+netstandard/SsoTokenUtils.cs (hunk 1 above)
Comment (awschristou) [reply sdk/src/Core/Amazon.Runtime/Credentials/Internal/_bcl45+netstandard/SsoTokenUtils.cs:112]: I think that knowing the system is working with clean data moving forward is better than the potential risk of leaving users in a scenario where they are unable to establish a connection, without manual intervention (manually deleting the cached tokens).

File: sdk/src/Core/Amazon.Runtime/Credentials/Internal/_bcl45+netstandard/SsoTokenUtils.cs [hunk 2]
```diff
@@ -49,7 +49,7 @@ private static class JsonPropertyNames
         public static bool IsExpired(this SsoToken token)
         {
             if (null == token)
-                throw new ArgumentNullException(nameof(token));
+                return true;
```
[the remark this answers is not in the document]
Comment (peterrsongg) [reply sdk/src/Core/Amazon.Runtime/Credentials/Internal/_bcl45+netstandard/SsoTokenUtils.cs:52]: You're right I shouldn't be changing this, I'll fix it. Got lost in the shuffle between different branches.

File: sdk/src/Core/Amazon.Runtime/Credentials/Internal/_bcl45+netstandard/SsoTokenUtils.cs [hunk 3]
```diff
@@ -99,13 +99,17 @@ public static string AsJson(this SsoToken token)
...
         /// <summary>
         /// This determines whether the <seealso cref="SsoToken.RegistrationExpiresAt"/> field is 5 minutes within expiration
+        /// Existing cached tokens, or tokens written out by other systems may contain null values for the RegistrationExpiresAt field.
+        /// In these cases we treat the client registration as expired and generate a new token.
         /// </summary>
         /// <param name="token">The sso token</param>
         /// <returns>This returns true if <seealso cref="SsoToken.RegistrationExpiresAt"/> is within 5 minutes of expiration. False otherwise</returns>
         public static bool RegisteredClientExpired(this SsoToken token)
         {
             if (null == token)
```
[the remark this answers is not in the document]
Comment (peterrsongg) [reply sdk/src/Core/Amazon.Runtime/Credentials/Internal/_bcl45+netstandard/SsoTokenUtils.cs:109]: nope, 52 was a mistake

**Comment by peterrsongg** [comment]
closing this for now to reopen a fresher PR at a later date.
